### PR TITLE
[NFC][flang][OpenMP] Split `DataSharing` and `Clause` processors

### PR DESCRIPTION
**Pull request by ergawy**
This started as an experiment to reduce the compilation time of iterating over `Lower/OpenMP.cpp` a bit since it is too slow at the moment. Trying to do that, I split the `DataSharingProcessor`, `ReductionProcessor`, and `ClauseProcessor` into their own files and extracted some shared code into a util file. All of these new `.h/.cpp` files as well as `OpenMP.cpp` are now under a `Lower/OpenMP/` directory.

This resulted is a slightly better organization of the OpenMP lowering code and hence opening this NFC.

As for the compilation time, this unfortunately does not affect it much (it shaves off a few seconds of `OpenMP.cpp` compilation) since from what I learned the bottleneck is in `DirectivesCommon.h` and `PFTBuilder.h` which both consume a lot of time in template instantiation it seems.

### CODE REVIEW

**Comment by kparzysz** [comment]
This makes sense to me.  I'd create a subdirectory `OpenMP` and move these files in there (and rename `OpenMPUtils` to just `Utils`).  What about `ReductionProcessor`?

**Comment by ergawy** [comment]
> This makes sense to me. I'd create a subdirectory `OpenMP` and move these files in there (and rename `OpenMPUtils` to just `Utils`). What about `ReductionProcessor`?

Thanks for taking a look @kparzysz. Moved to sub-dir and also split `ReductionProcessor`.

**Comment by ergawy** [comment]
Ping 🔔! Can you please have a look? 🙏

**Review by skatrak — COMMENTED** [review]
Thank you Kareem for this work, I have a couple of small comments but I think this refactoring is worth doing.

**Comment by kiranchandramohan** [comment]
Nice work. Would we be able to move OpenMP.cpp also to the OpenMP directory?

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,930 @@
...
+#include "flang/Parser/tools.h"
+#include "flang/Semantics/tools.h"
+
+namespace Fortran {
+namespace lower {
+namespace omp {
+
+//===----------------------------------------------------------------------===//
+// Common helper functions
+//===----------------------------------------------------------------------===//
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp]: Moved to Utils.h, so I think there is no need for the comment there.

**Comment by ergawy** [comment]
> Nice work. Would we be able to move OpenMP.cpp also to the OpenMP directory?

We can do that. Done!

**Review by kiranchandramohan — APPROVED** [review]
This will bring some short term pain for all the patches in flight. On our side particularly this will impact (refactoring cost) for the array reduction work. But overall this is a good patch and it LGTM. Please wait for other reviewers.

Please update the summary and title to include all the changes.

**Review by kparzysz — APPROVED** [review]
LGTM, thanks!

**Pull request merged**